exercises/grep/README.md

```diff
@@ -0,0 +1,95 @@
+# Grep
+
+Search a file for lines matching a regular expression pattern. Return the line
+number and contents of each matching line.
+
+The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command can be used to search for lines in one or more files
+that match a user-provided search query (known as the *pattern*).
+
+The `grep` command takes three arguments:
+
+1. The pattern used to match lines in a file.
+2. Zero or more flags to customize the matching behavior.
+3. One or more files in which to search for matching lines.
+
+Your task is to implement the `grep` function, which should read the contents
+of the specified files, find the lines that match the specified pattern
+and then output those lines as a single string. Note that the lines should
```

Comment:
Do we need them to output or return the lines as a single string?  (We want to be careful here, as output such as to the screen or to a file, as opposed to the return of a method...) The tests as they are written right now does not support testing for output.

Reply:
That's a great question. The readme (originally from [problem-specifications](https://github.com/exercism/problem-specifications/blob/master/exercises/grep/description.md)) occasionally mention "output", but all the implementations I saw tested solutions using return values. I'm not sure how we would reliably test this otherwise.

There's some discussion around this [here](https://github.com/exercism/ruby/issues/946#issuecomment-471363505).

Reply:
> Not sure how we could reliably test this otherwise

Minitest has an assertion for output... but it wouldn't be hard even without it, as we have access to `STDOUT` and friends regardless.  Perhaps unsurprisingly, the assertion is `assert_output` and `must_output`.

Reply:
I think setting some running code with a "library guard" would be a great inclusion to the exercise.  It is a simple thing to teach (and I already mention it as a mentor when people do include running code, or I need to show a bug, I will explain why I wrote it like I did in the code, so they can copy and paste it directly in the library file).

Reply:
@kotp didn't know about `assert_output` and `must_output`, thanks!

> I think setting some running code with a "library guard" would be a great inclusion to the exercise.

Could you help me understand this better, please? Maybe share an instance of one of these mentoring sessions so I can see what you're referring to?

Reply:
```diff
- and then output those lines as a single string
+ and then output those lines.

Reply:
The readme comes from problem-specifications - maybe it's worth opening an issue there so the language is clarified?